tests/components/blebox/test_light.py

```diff
@@ -425,6 +453,66 @@ def apply_color(value, color_value):
     assert state.attributes[ATTR_RGBW_COLOR] == (0xC1, 0xD2, 0xF3, 0xC7)
 
 
+# note: it is a good idea to test wide range of values including those
+# outside of the operating range to make sure that bounds are properly
+# corrected
+@pytest.mark.parametrize("temp_requested", [1, 100, 150, 200, 300, 400])
+async def test_wlightbox_on_temp(
+    wlightbox_ctx2, hass: HomeAssistant, temp_requested: int
+) -> None:
+    """Test light on with temperature change."""
+
+    feature_mock, entity_id = wlightbox_ctx2
+
+    # note: testing color temperature modification is tricky as neither
+    # blebox_uniapi nor blebox devices operate on color_temperature concept.
+    # Instead, it uses a mixture of brightness and temperature value encoded
+    # with known formula that is exposed by return_color_temp_with_brightness()
+    # method. To make testing simpler let's hijack the requested value
+    # with transient nonlocal value.
+    transient_temp: int = -1
+
+    def return_color_temp_with_brightness(value, brightness):
+        nonlocal transient_temp
+        transient_temp = value
+        return [0x00, 0x39, 0xB0, 0xFF]
+
+    def initial_update():
+        feature_mock.is_on = False
+
+    feature_mock.color_temp = 128
+    feature_mock.async_update = AsyncMock(side_effect=initial_update)
+    feature_mock.return_color_temp_with_brightness = return_color_temp_with_brightness
+
+    await async_setup_entity(hass, entity_id)
+    feature_mock.async_update = AsyncMock()
+
+    state = hass.states.get(entity_id)
+    assert state.state == STATE_OFF
+
+    def turn_on(value):
+        feature_mock.is_on = True
+        assert value == [0x00, 0x39, 0xB0, 0xFF]
+        feature_mock.color_temp = transient_temp
+
+    feature_mock.async_on = AsyncMock(side_effect=turn_on)
```

Comment:
the return_value of async_on isn't really used. Instead of having the mock replicate the expectations of the light, I think you could rather check if the mock is called the right way. Otherwise you have 2 similarish logics to replicate and that is bad for maintainability

Reply:
As in, the update of the entity probably is tested in another test, so if you have tests that verify that your update works, you only need to check if you called the API the right way.

Reply:
I agree that the test was a bit bloated so I have streamlined it a little bit and removed everything that could be removed. Nevertheless I would prefer to keep the test flow going through the async_on instead of inspecting the mock call args. The reasons for this are:
- modifying temperature goes through the the "turn on" logic and that's the reason why it was so hard to get it right in the first place
- the problem with temperature is that blebox_uniapi (and the underlying devices) use completely different value scale so it has to be mapped in both directions (see one of my comments in the test)
- after my first attempt (before I had it unit-tested) this thing was "almost working" but the value read from the device after the update was always off by one step. The test flow going through ordinary async_on allows me to test not only that we call the device API in expected way but also that given assumption of its behaviour (mocked `turn_on`) we can comprehend values given in the correct way.

And regarding maintainability of these test, I agree: it is already painful.

Reply:
But you can patch the mock right? `feature_mock.is_on = True` - assert the current state is off - execute the action call - assert the state is on - assert that `turn_on` is called the right way?

Reply:
You mean this?

```diff
diff --git a/tests/components/blebox/test_light.py b/tests/components/blebox/test_light.py
index 618d51a0ad2..f6c1d84148b 100644
--- a/tests/components/blebox/test_light.py
+++ b/tests/components/blebox/test_light.py
@@ -477,12 +477,9 @@ async def test_wlightbox_on_temp(
         transient_temp = value
         return [0x00, 0x39, 0xB0, 0xFF]
 
-    def turn_on(_):
-        feature_mock.is_on = True
-        feature_mock.color_temp = transient_temp
-
     feature_mock.return_color_temp_with_brightness = return_color_temp_with_brightness
-    feature_mock.async_on = AsyncMock(side_effect=turn_on)
+    feature_mock.is_on = True
+    feature_mock.color_temp = transient_temp
 
     await async_setup_entity(hass, entity_id)
     await hass.services.async_call(
```